src/main/java/org/folio/qm/controller/interceptor/LoggingRequestFilter.java

```diff
@@ -0,0 +1,122 @@
+package org.folio.qm.controller.interceptor;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.time.Instant;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import lombok.extern.log4j.Log4j2;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.GenericFilterBean;
+import org.springframework.web.util.ContentCachingRequestWrapper;
+import org.springframework.web.util.ContentCachingResponseWrapper;
+
+import org.folio.spring.integration.XOkapiHeaders;
+
+@Log4j2
+@Component
+@ConditionalOnProperty(
+  prefix = "folio.logging.request",
+  name = "enabled"
+)
+public class LoggingRequestFilter extends GenericFilterBean {
+
+  private static final String START_TIME_ATTR = "startTime";
+
+  private final Level level;
+
+  public LoggingRequestFilter(@Value("${folio.logging.request.level: FULL}") Level level) {
+    this.level = level;
+  }
+
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+    throws IOException, ServletException {
+
+    filterWrapped(wrapRequest(request), wrapResponse(response), chain);
+  }
+
+  private void filterWrapped(ContentCachingRequestWrapper request, ContentCachingResponseWrapper response,
+                             FilterChain chain) throws ServletException, IOException {
+    filterBefore(request, response);
+    chain.doFilter(request, response);
+    filterAfter(request, response);
+    response.copyBodyToResponse();
+  }
+
+  private void filterBefore(ContentCachingRequestWrapper request, ContentCachingResponseWrapper response)
+    throws UnsupportedEncodingException {
+    request.setAttribute(START_TIME_ATTR, Instant.now().toEpochMilli());
+
+    var requestId = getRequestId(request);
+
+    log.info("[{}] ---> {} {} {}",
+      requestId,
+      request.getMethod(),
+      request.getRequestURI(),
+      request.getQueryString()
+    );
+
+    if (level.ordinal() >= Level.HEADERS.ordinal()) {
```

Comment:
It's better to take into account the level of logger also to avoid unnecessary manipulations with data:
in this case verify that log level is `info` or below

Reply:
Do check for log level before all filter logic